state/cluster/badger/mutator.go

```diff
@@ -218,3 +205,74 @@ func (m *MutableState) Extend(block *cluster.Block) error {
 	}
 	return nil
 }
+
+// checkDupeTransactionsInUnfinalizedAncestry checks for duplicate transactions in the un-finalized
+// ancestry of the given block, and returns a list of all duplicates if there are any.
+func (m *MutableState) checkDupeTransactionsInUnfinalizedAncestry(block *cluster.Block, includedTransactions map[flow.Identifier]struct{}, finalHeight uint64) ([]flow.Identifier, error) {
+
+	var duplicateTxIDs []flow.Identifier
+	err := fork.TraverseBackward(m.headers, block.Header.ParentID, func(ancestor *flow.Header) error {
+		payload, err := m.payloads.ByBlockID(ancestor.ID())
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor payload: %w", err)
+		}
+
+		for _, tx := range payload.Collection.Transactions {
+			txID := tx.ID()
+			_, duplicated := includedTransactions[txID]
+			if duplicated {
+				duplicateTxIDs = append(duplicateTxIDs, txID)
+			}
+		}
+		return nil
+	}, fork.ExcludingHeight(finalHeight))
+
+	return duplicateTxIDs, err
```

Comment:
`checkDupeTransactionsInUnfinalizedAncestry` has duplicated code with `populateUnfinalizedAncestryLookup`. 
This is how you can extract the common part:

```
func iterateThroughTransactionsInUnfinalizedAcenstery(clusterHeaders, payloads, parentID, finalHeight, reducer func(blockHeight, txID)) error {
  return fork.TraverseBackward(clusterHeaders, parentID, func(ancestor *flow.Header) error {
		payload, err :=payloads.ByBlockID(ancestor.ID())
		if err != nil {
			return fmt.Errorf("could not retrieve ancestor payload: %w", err)
		}

		for _, tx := range payload.Collection.Transactions {
			reducer(ancester.Height, tx.ID())
		}
		return nil
	}, fork.ExcludingHeight(finalHeight))
}

// for builder
iterateThroughTransactionsInUnfinalizedAcenstery(clusterHeaders, payloads, finalHeight, func(blockHeight, txID) {
  lookup.addUnfinalizedAncestor(txID)
  limiter.addAncestor(blockHeight)
})

// for mutator
iterateThroughTransactionsInUnfinalizedAcenstery(clusterHeaders, payloads, finalHeight, func(blockHeight, txID) {
  _, duplicated := includedTransaction[txID]
  if duplicated {
    duplicatedTx = append(duplicatedTx, txID)
  }
})
```

same for `checkDupeTransactionsInFinalizedAncestry`

Reply:
I didn't think it was worth doing this here, because:
* the common logic is only a few lines of code, and the only two usages are in different packages
* if we implement the [ID-only retrieval lookup for the builder](https://github.com/onflow/flow-go/pull/2385#discussion_r871884647), which I think we should, there is little re-factorable common logic left